text/0000-ghost-busting.md

```diff
@@ -0,0 +1,738 @@
+    let foo: Foo<u8, ()> = Foo {
+        bar: 1u8,
+        baz: 2usize,
+    };
+}
+```
+
+Note in particular that we didn't have to, and indeed can't, set a field
+`marker: PhantomData` in the struct literals above.
+
+## `phantom` fields
+
+Given the following type `MyVec<T>`, `T` will be covariant but also not counted
+as logically owned by the drop checker. Therefore, `MyVec<T>` is not sound.
+
+```rust
+struct MyVec<T> {
+    data: *const T,
+    len: usize,
+    cap: usize,
+}
+```
+
+Before this RFC, this was solved by adding a marker field with the type
+`PhantomData<T>` as in:
+
+```rust
+use std::marker::PhantomData;
+
+struct MyVec<T> {
+    marker: PhantomData<T>,
+    data: *const T,
+    len: usize,
+    cap: usize,
+}
+```
+
+This made `T` logically owned and covariant. With this RFC, you will
+instead write:
+
+```rust
+struct MyVec<T> {
+    phantom T,
+    data: *const T,
+    len: usize,
+    cap: usize,
+}
+```
+
+Another example of using `phantom` fields which was discussed in the
+[motivation] is:
+
+```rust
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct Id<A: ?Sized, B: ?Sized> {
+    // fn(A) -> A makes A invariant:
+    phantom (fn(A) -> A, fn(B) -> B)
+}
+
+impl<A: ?Sized> Id<A, A> { pub const REFL: Self = Id; }
+
+impl<A: ?Sized, B: ?Sized> Id<A, B> {
+    pub fn cast(self, value: A) -> B where A: Sized, B: Sized {
+        unsafe {
+            let cast_value = mem::transmute_copy(&value);
+            mem::forget(value);
+            cast_value
+        }
+    }
+}
+```
+
+In this particular example, and to maintain type and memory safety, we need to
+change the default covariant variance to invariant because we would otherwise
+introduce the ability to transmute lifetimes in safe Rust with:
+
+```rust
+fn transmute_lifetime<'a, 'b, T: 'a + 'b>(r: &'a T) -> &'b T {
+    Id::REFL.cast(r)
+}
+```
+
+### Not nameable
+
+It is important to note that the introduced `phantom` fields are truly
+phantom, unlike with `marker: PhantomData<T>`, since you can't reference
+or name a `phantom` field either during construction in a literal expression
+or with field access notation `expr.field`.
+
+A consequence of this is that given the following type, you can coerce `Foo`
+into a function pointer of type `fn(u8) -> Foo<u8, SomeType>`:
+
+```rust
+struct Foo<X, Y>(X, phantom Y);
+```
+
+### Visibility
+
+[visibility modifiers]: https://doc.rust-lang.org/nightly/reference/visibility-and-privacy.html
+
+While `phantom` fields are not nameable, you can put [visibility modifiers] on
+them or none, in which case the default private visibility is used. Visibility
+on a `phantom` field has the standard effect that the type it is in can't be
+constructed unless the "field" is visible in the module in question.
+
+This property is crucial because if `phantom` fields always were public, then
+the a value of type `Id<A, B>` where `A != B` could be constructed with `Id {}`
```

Comment:
This specific instance could (and should) be fixed using the [non_exhaustive](https://doc.rust-lang.org/nightly/unstable-book/language-features/non-exhaustive.html) attribute. Do you have a better use case for privacy in phantom fields?

Reply:
The documentation of `#[non_exhaustive]` says:

> Structs marked as non_exhaustive will not be able to be created normally outside of the defining crate.

Essentially, this is the same as `pub(crate)`.

Allowing for more fine grained visibility as `pub(super)` is more general and lets you control visibility on level not permitted by `#[non_exhaustive]`. I also believe that `pub(crate)` is a lot more clear than `#[non_exhaustive]` wrt. intent and what it means. I think that `#[non_exhaustive]` should mostly be used for enums and not for structs.

Reply:
It's not like you couldn't work around it by writing

```rust
struct Id<A, B> { phantom A, phantom B, _priv: () }
```

It *really* makes no sense for phantom fields to be subject to visibility checks. You can always add a unit field, and if that's not acceptable, then the problem can be fixed in a more general way.

Reply:
The idea of using a _priv field is discussed in the alternatives section. You say that there is no reason for phantom fields to be subject to visibility, but a _priv field will need to be passed in struct literals, making things less ergonomic, so there are reasons to make phantom fields subject to visibility. It also increases uniformity of fields to do so.

Reply:
It seems to me that `phantom ()` is a more natural way to force a private field, than requiring it be named anyway.